### ISSUE TITLE
Testing how to use finite differences for testing

### ISSUE THREAD

**Issue by facusapienza21** [ODINN-SciML/ODINN.jl#305]
I was testing and checking the code for the general inversions (both A and D in my last PR #298 ) and I started wondering how much of the error between the adjoint code and the reference gradient is coming form the gradient calculation and how much from the finite difference code. As it is right now, out tests try to find the best match for all coordinates all at once for different choices of the stepsize epsilon. However, I see three issues with this: 
- The epsilon is the same for each coordinate of the gradient, which it should not be necessary the case
- It is unclear that the lowest error corresponds to the best choice of epsilon (maybe the error is just smaller by chance, not because the gradient is actually better).
- In principle, we are just testing for the forward difference scheme, which is just order O(epsilon) and maybe we want a more precise stencil. 

This makes me thing that maybe we can just use `FiniteDifferences.jl` to automatically pick the stepsize (potentially, custom per coordinate) and allow for a higher order estimation of the gradient. I gave this a try in #298 by adding the following code in the test: 
```julia 
  dθ_FD, = FiniteDifferences.grad(
      central_fdm(7, 1),
      _θ -> loss_function(_θ, (simulation)),
      θ
  )
  ratio_FD, angle_FD, relerr_FD = stats_err_arrays(dθ, dθ_FD)
  @test ratio_FD < thres_ratio
  @test angle_FD < thres_angle
  @test relerr_FD < thres_relerr
  printVecScientific("ratio  = ", [ratio_FD], thres_ratio)
  printVecScientific("angle  = ", [angle_FD], thres_angle)
  printVecScientific("relerr = ", [relerr_FD], thres_relerr)
```
Notice here the finite differences is computed with a center scheme of order 7th (which can be further customize). The output of this code compared with the one using finite differences for the test on `ContinuousAdjoint(DiscreteVJP())` with target `A` gives the following errors: 
![Image](https://github.com/user-attachments/assets/96278027-cffb-46c2-ad7a-74243e75cea7)
All tests passing, we can see that the error using `FiniteDifferences.jl` is smaller than the one we currenly have by picking the best among all epsilons (again, best assuming each coordinate uses the same stepsize). For other choices of adjoints/VJPs in the tests, the errors fluctuate. This is also a good call to think what are good thresholds here. 

@albangossard I would love to hear your thought on this! We can keep both tests, or just use the ones with finite differences. I think it will be better to automatize this with FiniteDifferences.

**Comment by albangossard** [comment]
Thanks for this initiative @facusapienza21! I definitely agree that it's better to use `FiniteDifferences.jl` than recoding everything ourselves. A few comments just for general discussion to follow up on what you said:

> The epsilon is the same for each coordinate of the gradient, which it should not be necessary the case

In the specific setting where all components are of the same "nature" like when we are differentiating the SIA with respect to H, this makes sense. But I agree that since we moved to the differentiation of the neural network, this doesn't hold as each component as a very different impact on the loss function.

> It is unclear that the lowest error corresponds to the best choice of epsilon (maybe the error is just smaller by chance, not because the gradient is actually better).

Since we are computing the error, the angle and the ratio of the norm, the best epsilon is the one that allows decreasing these three quantities. In an exact arithmetic setting, the smaller the epsilon the lower the error, but since there are numerical errors, there is a trade-off between gradient approximation and numerical error. This translates in the error having a V shape where the error first starts to decrease as epsilon goes towards zero, but at some point the numerical error dominates and the error starts to increase. I have some examples in [this notebook](https://github.com/albangossard/TP-AN-optim-3MIC/blob/master/notebooks/TP1_Correction.ipynb) in the cell n°5. From my experience this behavior always appears.

> We can keep both tests, or just use the ones with finite differences. I think it will be better to automatize this with FiniteDifferences.

I completely agree, there is a package that does even more than what we have implemented, so we should move to that. Just one question: what is the computational time of performing the test with `FiniteDifferences.jl` in comparison to our naive implementation? Even if it's slower, being able to change the order of the numerical scheme is worth it! That's just a question out of curiosity.

**Comment by albangossard** [comment]
> Yes, I am familiar with the V shape (we include a version of it in the review paper comparing also with AD). But in those cases you are comparing the real error, meaning "finite diff vs real derivative". However, we are comparing "custom adjoints vs finite differences". The epsilon corresponding to the minimum of the latest does not necessarily is the same as the minumum of the first one. For testing, we want the epsilon that minimizes the first to compute finite differences and then use this one to report the second. Right now we are just minimizing the second, which is not necessarily true. I agree the V shape effect is always there, reason why it is important to be careful how do we use finite differences for testing.

I understand your point now, thank you for this clarification. Yes I agree that our methodology is not correct and that we should report the optimal epsilon used in the real derivative test to the custom adjoint tests.
If you can find a way to do that without manually hardcoding the epsilon values in the tests, feel free to give it a try. However, I think this is a low priority task as the purpose of the tests is just to ensure that we don't break the adjoints by mistake.

**Comment by facusapienza21** [comment]
Ok, I run some simple benchmark with different finite difference codes: 
```julia
res1 = @benchmark test_grad_finite_diff(ContinuousAdjoint(VJP_method=DiscreteVJP()); thres=[1.0, 1.0, 1.0], finite_difference_method = :Manual)
res2 = @benchmark test_grad_finite_diff(ContinuousAdjoint(VJP_method=DiscreteVJP()); thres=[1.0, 1.0, 1.0], finite_difference_method = :FiniteDifferences, finite_difference_order = 2)
res3 = @benchmark test_grad_finite_diff(ContinuousAdjoint(VJP_method=DiscreteVJP()); thres=[1.0, 1.0, 1.0], finite_difference_method = :FiniteDifferences, finite_difference_order = 3)
res4 = @benchmark test_grad_finite_diff(ContinuousAdjoint(VJP_method=DiscreteVJP()); thres=[1.0, 1.0, 1.0], finite_difference_method = :FiniteDifferences, finite_difference_order = 5)
res5 = @benchmark test_grad_finite_diff(ContinuousAdjoint(VJP_method=DiscreteVJP()); thres=[1.0, 1.0, 1.0], finite_difference_method = :FiniteDifferences, finite_difference_order = 7)
```
Notice that here the functions have the `finite_difference_method` which indicates if we use FiniteDifferences.jl or the manual implementation, and for the former one extra integer parameter with the order of the approximation. These are the results: 
- Manual Implementation
```
BenchmarkTools.Trial: 3 samples with 1 evaluation per sample.
 Range (min … max):  40.361 s …   41.841 s  ┊ GC (min … max): 20.72% … 20.81%
 Time  (median):     41.471 s               ┊ GC (median):    20.17%
 Time  (mean ± σ):   41.224 s ± 770.091 ms  ┊ GC (mean ± σ):  20.53% ±  0.41%

  █                                           █             █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  40.4 s          Histogram: frequency by time         41.8 s <

 Memory estimate: 306.06 GiB, allocs estimate: 10589091.
```
- `FiniteDifferences.jl` with `order = 2`
```
BenchmarkTools.Trial: 4 samples with 1 evaluation per sample.
 Range (min … max):  37.705 s …   38.319 s  ┊ GC (min … max): 18.24% … 19.32%
 Time  (median):     37.871 s               ┊ GC (median):    18.44%
 Time  (mean ± σ):   37.941 s ± 264.412 ms  ┊ GC (mean ± σ):  18.57% ±  0.58%

  █            █    █                                       █  
  █▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  37.7 s          Histogram: frequency by time         38.3 s <

 Memory estimate: 279.20 GiB, allocs estimate: 9726247.

```
- `FiniteDifferences.jl` with `order = 3`
```
BenchmarkTools.Trial: 2 samples with 1 evaluation per sample.
 Range (min … max):  73.552 s …   74.113 s  ┊ GC (min … max): 57.73% … 66.55%
 Time  (median):     73.832 s               ┊ GC (median):    62.16%
 Time  (mean ± σ):   73.832 s ± 396.575 ms  ┊ GC (mean ± σ):  62.16% ±  6.23%

  █                                                         █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  73.6 s          Histogram: frequency by time         74.1 s <

 Memory estimate: 369.80 GiB, allocs estimate: 12631295.
```
- `FiniteDifferences.jl` with `order = 5`
```
BenchmarkTools.Trial: 2 samples with 1 evaluation per sample.
 Range (min … max):  101.086 s …  101.905 s  ┊ GC (min … max): 67.00% … 67.34%
 Time  (median):     101.496 s               ┊ GC (median):    67.17%
 Time  (mean ± σ):   101.496 s ± 579.285 ms  ┊ GC (mean ± σ):  67.17% ±  0.24%

  █                                                          █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  101 s           Histogram: frequency by time           102 s <

 Memory estimate: 539.28 GiB, allocs estimate: 18067295.
```
- `FiniteDifferences.jl` with `order = 7`
```
BenchmarkTools.Trial: 1 sample with 1 evaluation per sample.
 Single result which took 137.121 s (67.51% GC) to evaluate,
 with a memory estimate of 714.47 GiB, over 23686237 allocations.
```
The running time of course increases with the order, but I think in general they are all quite fine and I will probably recommend keeping `order = 5` for a good estimation of the derivative without going crazy.

**Comment by albangossard** [comment]
Thank you for having taken the time to make this benchmark! This is interesting, I would have expected `FiniteDifferences.jl` to be slower than it is.

**Comment by facusapienza21** [comment]
Closed in #298
